File: src/content-parser/league_parser.js
const ContentParser = require('./content-parser');
const {
  EndTokenNotFoundParserException,
  ParserException,
  StartTokenNotFoundParserException,
  TagNotFoundParserException,
  TeamCatalogParserException,
} = require('../exceptions/exceptions');
const { Match } = require('../schemas/dtos');


class TeamLeagueContentParser extends ContentParser {
  parse() {
    try {
      this.content = this.staticTextParser.parse('LEAGUE').value();
    } catch (ex) {
      console.log('Error occurred during parsing of the team catalog.')
      throw new TeamCatalogParserException(ex?.message);
    }

    const matches = [];
    let foundMatch = true;

    let parameter_name = null;

    while (foundMatch) {
      const data = {};

      for (const parameter of this.template.parameters) {
        try {
          const value = this.sequentialTextParser.parse(parameter.name).value({
            uppercase: true,
            chainFilter: this.filters[parameter.name],
          });
          parameter_name = parameter.name;

          try {
            data[parameter.name.toLowerCase()] = value;
          } catch (error) {
            console.log('ERROR:', error)
            continue;
          }
        } catch (ex) {
          if (
            ex instanceof StartTokenNotFoundParserException ||
            ex instanceof EndTokenNotFoundParserException ||
            ex instanceof TagNotFoundParserException
          ) {
            foundMatch = false;
            break;
          } else {
            console.log(ex.message);
            throw new ParserException(ex.message);
          }
        }
      }

      if (data && foundMatch) {
        const match = new Match(data);
        matches.push(match);
      }
    }
    return matches;
  }
}

module.exports = {TeamLeagueContentParser};

File: src/content_parser/league_parser.py
import logging

from content_parser.content_parser import ContentParser
from exceptions.parser_exceptions import (
    EndTokenNotFoundParserException,
    ParserException,
    StartTokenNotFoundParserException,
    TagNotFoundParserException,
    TeamCatalogParserException,
)
from schemas.dtos import Match

logger = logging.getLogger(__name__)


class TeamLeagueContentParser(ContentParser):
    """Content parser for team league information.

    This class extends the ContentParser abstract base class and provides specific parsing logic
    for extracting team league information from the content."""

    def parse(self):
        try:
            self.content = self.text_parser_static.parse('LEAGUE').value()
        except ParserException as ex:
            logger.error('Error occurred during parsing of the team catalog.')
            raise TeamCatalogParserException from ex

        matches = []

        found_match = True

        while found_match:
            match = Match.construct()

            for parameter in self.template.parameters:
                try:
                    value = self.text_parser_sequential.parse(parameter.name).value(
                        chain_filter=self.filters.get(parameter.name)
                    )
                    try:
                        setattr(match, parameter.name.lower(), value)
                    except (ValueError, AttributeError):
                        continue
                except (
                    StartTokenNotFoundParserException,
                    EndTokenNotFoundParserException,
                    TagNotFoundParserException,
                ):
                    logger.warning('No more matching results in recursive search')
                    found_match = False
                    break

            if match and found_match:
                logger.info(f'Valid match found: {match.home_team}')
                matches.append(match)

        logger.info(f'Returning {len(matches)} valid matches')
        return matches
